include/picongpu/fields/background/templates/twtstight/TWTSTight.tpp

```diff
@@ -0,0 +1,366 @@
+                    return (*this)(cellIdx, currentStep)[T_component];
+            }
+
+            template<typename T_Field>
+            HDINLINE std::tuple<
+                float_T,
+                float_T,
+                float_T,
+                float_T,
+                float_T,
+                float_T,
+                float_T,
+                float_T,
+                float_T,
+                float_T,
+                float_T>
+            TWTSTight<T_Field>::defineBasicHelperVariables() const
+            {
+                /* If phi < 0 the formulas below are not directly applicable.
+                 * Instead phi is taken positive, but the entire pulse rotated by 180 deg around the
+                 * y-axis of the coordinate system in this function.
+                 */
+                auto const absPhi = float_T(math::abs(phi));
+
+                /* Propagation speed of overlap normalized to the speed of light [Default: beta0=1.0] */
+                auto const beta0 = float_T(beta_0);
+
+                float_T sinPhi;
+                float_T cosPhi;
+                math::sincos(absPhi, sinPhi, cosPhi);
+                float_T const tanAlpha = (float_T(1.0) - beta0 * cosPhi) / (beta0 * sinPhi);
+
+                auto const cspeed = float_T(sim.si.getSpeedOfLight() / sim.unit.speed());
+                auto const lambda0 = float_T(wavelength_SI / sim.unit.length());
+                float_T const omega0 = float_T(2.0 * PI) * cspeed / lambda0;
+                /* factor 2  in tauG arises from definition convention in laser formula */
+                auto const tauG = float_T(pulselength_SI * 2.0 / sim.unit.time());
+                /* w0 is wx here --> w0 could be replaced by wx */
+                auto const w0 = float_T(w_x_SI / sim.unit.length());
+                auto const k = float_T(2.0 * PI / lambda0);
+
+                return std::make_tuple(absPhi, sinPhi, cosPhi, beta0, tanAlpha, cspeed, lambda0, omega0, tauG, w0, k);
+            }
+
+            template<typename T_Field>
+            HDINLINE std::tuple<float_T, float_T, float_T, float_T> TWTSTight<T_Field>::defineMinimalCoordinates(
+                float3_64 const& pos,
+                float_64 const& time) const
+            {
+                /* In order to calculate in single-precision and in order to account for errors in
+                 * the approximations far from the coordinate origin, we use the wavelength-periodicity and
+                 * the known propagation direction for realizing the laser pulse using relative coordinates
+                 * (i.e. from a finite coordinate range) only. All these quantities have to be calculated
+                 * in double precision.
+                 */
+                float_64 const deltaT = wavelength_SI / sim.si.getSpeedOfLight()
+                    / (1.0 - beta_0 * math::cos(precisionCast<float_64>(phi)));
+                float_64 const deltaY = beta_0 * sim.si.getSpeedOfLight() * deltaT;
```

Comment:
Cluttering the class with const data members that are consecutively initialized in initializer lists is a nightmare because I need to keep track of the order of member declarations. Nested structs clutter the code when it is used. Is there a syntax-suger solution to this problem?